README.md
# koa-respond

[![npm](https://img.shields.io/npm/v/koa-respond.svg)](https://www.npmjs.com/package/koa-respond)
[![dependency Status](https://img.shields.io/david/jeffijoe/koa-respond.svg)](https://david-dm.org/jeffijoe/koa-respond)
[![devDependency Status](https://img.shields.io/david/dev/jeffijoe/koa-respond.svg)](https://david-dm.org/jeffijoe/koa-respond)
[![Build Status](https://img.shields.io/travis/jeffijoe/koa-respond.svg)](https://travis-ci.org/jeffijoe/koa-respond)
[![Coveralls](https://img.shields.io/coveralls/jeffijoe/koa-respond.svg)](https://coveralls.io/github/jeffijoe/koa-respond)
[![Code Climate](https://img.shields.io/codeclimate/github/jeffijoe/koa-respond.svg)](https://codeclimate.com/github/jeffijoe/koa-respond)
[![npm](https://img.shields.io/npm/dt/koa-respond.svg)](https://www.npmjs.com/package/koa-respond)
[![license](https://img.shields.io/npm/l/koa-respond.svg)](https://github.com/jeffijoe/koa-respond/blob/master/LICENSE.md)
[![node](https://img.shields.io/node/v/koa-respond.svg)](https://www.npmjs.com/package/koa-respond)
[![JavaScript Style Guide](https://img.shields.io/badge/code%20style-standard-brightgreen.svg)](http://standardjs.com/)
<a href="https://communityinviter.com/apps/koa-js/koajs" rel="KoaJs Slack Community">![KoaJs Slack](https://img.shields.io/badge/Koa.Js-Slack%20Channel-Slack.svg?longCache=true&style=for-the-badge)</a>


Middleware for Koa that adds useful methods to the Koa context.

## Installation

```bash
npm install --save koa-respond
```

## Usage

```js
// Install it
const respond = require('koa-respond');

// For Koa v2 - if you are looking for v1, scroll to the bottom.
app.use(respond());

// Use it
app.use((ctx) => {
  // Sets status to 200 and the body to `{ id: 123, name: 'Dat Boi' }`
  ctx.ok({ id: 123, name: 'Dat Boi' });

  // Both of these set status to 404 and
  // the body to `{ message: 'Not found, boii' }`
  ctx.notFound('Not found, boii');
  ctx.notFound({ message: 'Not found, boii' });

  // And everyone's favorite..
  ctx.badRequest({ error: 'missing input' });

  // Or if you prefer to do it yourself..
  // Both of these send a HTTP 201 with a body
  // of `{ message: 'new beginnings!' }`
  ctx.send(201, 'new beginnings!');
  ctx.send(201, { message: 'new beginnings!' });
});
```

## Methods

All methods call the `send` method with the corresponding status code as well as the body. That means they support the same overloads as `send`:

* With a string; wraps it in an object with a `message` property. That means the following 2 calls do the same thing:

  ```js
  ctx.send(400, 'lol no');
  ctx.send(400, { message: 'lol no' });
  ```

* With an object; sends the object as JSON.

  ```js
  ctx.send(200, { id: 123, name: 'new entity' });
  ```

If you wish to disable the automatic wrapping of strings globally, you can instantiate `koa-respond` with `autoMessage: false`.

```js
app.use(respond({
  autoMessage: false
}))
```

All functions return the Koa context itself (chainable)

```js
ctx.ok().set({ 'X-Some-Header': 'awesome' })
```

All functions are also bound to the context. This means you can pass the function as a reference without having to bind it first.

```js
app.use((ctx) => somePromiseCall().then(ctx.ok))
```

### Available methods

* `ok` - `HTTP 200`
* `created` - `HTTP 201`
* `noContent` - `HTTP 204` - **always sends an empty response!**
* `badRequest` - `HTTP 400`
* `unauthorized` - `HTTP 401`
* `forbidden` - `HTTP 403`
* `notFound` - `HTTP 404`
* `locked` - `HTTP 423`
* `internalServerError` - `HTTP 500`
* `notImplemented` - `HTTP 501`

## Does this work for Koa 1?

**Not out of the box**, because it's time you move on to v2.

To use `koa-respond` in Koa v1, you need to patch the context yourself. This is what the v2 middleware does.

```js
const respond = require('koa-respond');

// Middleware to install koa-respond.
app.use(function *(next) {
  respond().patch(this);
  yield next;
});

// Now the methods are available.
app.use(function *() {
  this.ok({ id: 123, name: 'Bob' });
});
```

## Adding additional methods

If you feel like some methods are missing, you can add them yourself, like so:

```js
app.use(respond({
  statusMethods: {
    imATeapot: 418,
    enhanceYourCalm: 420
  }
}));

app.use((ctx) => {
  ctx.imATeapot('Hello, a Teapot I am.');
  ctx.enhanceYourCalm({ todo: 'blaze it' });
});
```

## Even more custom methods

If you just want to add shortcuts without adding an additional middleware, you can do that, too.

```js
app.use(respond({
  methods: {
    shizzle: (ctx, message) => {
      ctx.send(200, message + ', fo-shizzle');
    }
  }
}));

app.use((ctx) => {
  // HTTP 200 { message: 'Koa is the best, fo-shizzle' }
  ctx.shizzle('Koa is the best');
});
```

# Contributing

## `npm run` scripts

* `npm run test`: Runs tests once
* `npm run test-watch`: Runs tests in watch-mode
* `npm run lint`: Lints the code once
* `npm run lint-watch`: Lints the code in watch-mode
* `npm run cover`: Runs code coverage using `istanbul`
* `npm run coveralls`: Used by coveralls

# Author

Jeff Hansen - [@Jeffijoe](https://twitter.com/Jeffijoe)


CHANGES.md
# Changelog

- 2.1.0

  - Add `.locked()` (HTTP 423) and `.notImplemented()` (HTTP 501) ([#4](/pull/4) — [crobinson42](https://github.com/crobinson42))

- 2.0.0

  - Update packages and set minimum Node version to 8 (LTS at time of release).

- 1.0.1

  - Use `standard` code style.

- 1.0.0

  - **[BREAKING]**: renamed `ctx.respond` to `ctx.send` due to [a naming conflict in Koa](http://koajs.com/#ctx-respond).

- 0.3.0
  - Initial release


test/lib/respond.spec.js
'use strict'

const makeRespond = require('../../lib/respond')

const respond = makeRespond()
const respondNoAutomessage = makeRespond({
  autoMessage: false
})

describe('respond', function() {
  let ctx

  beforeEach(function() {
    ctx = {
      status: 404,
      body: 'unset'
    }
  })

  it('sets the status on the context', function() {
    respond(ctx, 200)
    ctx.status.should.equal(200)
    ctx.body.should.equal('unset')
  })

  it('returns the given context', function() {
    const result = respond(ctx, 400)
    result.should.equal(ctx)
  })

  describe('when payload is a string', function() {
    it('sets an object with a message property', function() {
      respond(ctx, 200, 'A fine body')
      ctx.status.should.equal(200)
      ctx.body.should.deep.equal({ message: 'A fine body' })
    })
  })

  describe('when payload is anything but a string', function() {
    it('passes it through as-is', function() {
      const obj = { id: 123, name: 'koa-respond' }
      respond(ctx, 200, obj)
      ctx.status.should.equal(200)
      ctx.body.should.deep.equal(obj)
    })
  })

  describe('when autoMessage = false', function() {
    it('does not wrap strings', function() {
      respondNoAutomessage(ctx, 200, 'Hello world')
      ctx.body.should.equal('Hello world')
    })
  })
})


test/lib/koa-respond.spec.js
'use strict'
const respond = require('../../lib/koa-respond')

describe('koa-respond', function() {
  let ctx

  function use(middleware) {
    middleware(ctx, () => Promise.resolve())
  }

  beforeEach(function() {
    ctx = {
      status: 404
    }
  })

  it('exists', function() {
    expect(respond).to.exist
  })

  it('adds a send method', function() {
    use(respond())
    ctx.send.should.be.a('function')
  })

  describe('ctx.respond', function() {
    it('sets the status and body', function() {
      use(respond())
      ctx.send(200, { id: 123 })
      ctx.status.should.equal(200)
      ctx.body.should.deep.equal({ id: 123 })
    })
  })

  describe('statusMethods option', function() {
    it('adds custom methods', function() {
      use(
        respond({
          statusMethods: {
            imATeapot: 418,
            nope: 503
          }
        })
      )

      ctx.imATeapot('funny')
      ctx.status.should.equal(418)
      ctx.body.should.deep.equal({ message: 'funny' })

      ctx.nope('offline')
      ctx.status.should.equal(503)
      ctx.body.should.deep.equal({ message: 'offline' })
    })
  })

  describe('methods option', function() {
    it('adds custom methods', function() {
      use(
        respond({
          methods: {
            caps: (input, message) => {
              input.should.equal(ctx)
              input.send(200, message.toUpperCase())
            }
          }
        })
      )

      ctx.caps('never gonna give you up')
      ctx.status.should.equal(200)
      ctx.body.message.should.equal('NEVER GONNA GIVE YOU UP')
    })
  })

  // Generate tests for the built-in status codes.
  for (const method in respond.statusCodeMap) {
    const code = respond.statusCodeMap[method]
    describe(`ctx.${method}`, function() {
      it(`sets the status to ${code}`, function() {
        use(respond())
        ctx[method]({ id: 123 })
        ctx.status.should.equal(code)
        if (code === 204) {
          expect(ctx.body).to.be.null
        } else {
          ctx.body.should.deep.equal({ id: 123 })
        }
      })
    })
  }
})


test/integration/integration.spec.js
'use strict'

const respond = require('../../lib/koa-respond')
const supertest = require('supertest')
const Koa = require('koa')
const KoaRouter = require('koa-router')

const app = new Koa()
const router = new KoaRouter()

app.use(respond())

for (const method in respond.statusCodeMap) {
  router.get(`/${method}`, ctx => {
    ctx[method]({ message: 'hello' })
  })
}

app.use(router.routes())
const request = supertest(app.callback())

for (const method in respond.statusCodeMap) {
  const code = respond.statusCodeMap[method]
  if (code === 204) continue
  describe(`ctx.${method}`, function() {
    it(`sets the code to ${code}`, function() {
      return request
        .get(`/${method}`)
        .expect(code)
        .expect({ message: 'hello' })
    })
  })
}

describe('ctx.noContent', function() {
  it('sets the code to 204 and has no content', function() {
    return request
      .get('/noContent')
      .expect(204)
      .end(res => {
        expect(res.body).to.be.null
      })
  })
})


lib/koa-respond.js
'use strict'
const makeRespond = require('./respond')

/**
 * Maps method names to status codes.
 * These are just the most common ones.
 *
 * @type {Object}
 */
const statusCodeMap = {
  ok: 200,
  created: 201,
  noContent: 204,
  badRequest: 400,
  unauthorized: 401,
  forbidden: 403,
  notFound: 404,
  locked: 423,
  internalServerError: 500,
  notImplemented: 501
}

/**
 * Makes the respond middleware. All options are optional.
 *
 * @param {object} opts
 * Options object.
 *
 * @param {object} opts.statusMethods
 * An object where keys maps to method names, and
 * values map to the status code.
 *
 * @return {Function}
 */
function makeRespondMiddleware(opts) {
  opts = Object.assign({}, opts)

  // Make the respond function.
  const respond = makeRespond(opts)

  /**
   * Installs the functions in the context.
   *
   * @param  {KoaContext} ctx
   */
  function patch(ctx) {
    const statusMethods = Object.assign({}, opts.statusMethods, statusCodeMap)
    ctx.send = respond.bind(ctx, ctx)

    // Bind status methods.
    for (const method in statusMethods) {
      const code = statusMethods[method]
      ctx[method] = respond.bind(ctx, ctx, code)
    }

    // Bind other methods
    const methods = Object.assign({}, opts.methods)
    for (const method in methods) {
      const fn = methods[method]
      ctx[method] = fn.bind(ctx, ctx)
    }

    return ctx
  }

  /**
   * The respond middleware adds the methods to the context.
   *
   * @param  {KoaContext} ctx
   */
  function respondMiddleware(ctx, next) {
    patch(ctx)
    return next()
  }

  // Tack on the patch method to allow Koa 1 users
  // to install it, too.
  respondMiddleware.patch = patch
  return respondMiddleware
}

module.exports = makeRespondMiddleware

// Expose these for the tests.
module.exports.statusCodeMap = statusCodeMap
